File: WebClient SDK/src/default/cards/krPermissionsMandatoryStoreExtension.ts
import { CardStoreExtension, ICardStoreExtensionContext } from 'tessa/cards/extensions';
import { UIContext, tryGetFromInfo } from 'tessa/ui';
import { systemKeyPrefix, CardTaskAction } from 'tessa/cards';
import { getTypedFieldValue, TypedField, DotNetType } from 'tessa/platform';
import { ICardModel } from 'tessa/ui/cards';
import { LocalizationManager } from 'tessa/localization';
import { DefaultFormTabWithTasksViewModel } from 'tessa/ui/cards/forms';
import { TaskViewModel } from 'tessa/ui/cards/tasks';

interface KrPermissionMandatoryRuleStorage {
  sectionId: guid;
  columnIds: guid[];
}

export class KrPermissionsMandatoryStoreExtension extends CardStoreExtension {

  public afterRequest(context: ICardStoreExtensionContext) {
    // tslint:disable-next-line:no-any
    let rules: any[] | null = null;
    if (!context.requestIsSuccessful
      && UIContext.current.cardEditor
      && context.response
      && (rules = tryGetFromInfo(context.response.info, systemKeyPrefix + 'FailedMandatoryRules'))
    ) {
      const cardModel = UIContext.current.cardEditor.cardModel!;
      const failRules: KrPermissionMandatoryRuleStorage[] = rules.map(x => {
        return {
          sectionId: getTypedFieldValue(x['SectionID']),
          columnIds: (x['ColumnIDs'] as Array<TypedField<DotNetType.Guid, guid>>).map(y => getTypedFieldValue(y))
        };
      });

      this.validateFromRules(cardModel, failRules);

      const requestTasks = context.request.card.tasks;
      const taskItems = cardModel.mainForm instanceof DefaultFormTabWithTasksViewModel ? cardModel.mainForm.tasks : null;
      if (requestTasks.length > 0
        && !!taskItems
        && taskItems.length > 0
      ) {
        for (let task of requestTasks) {
          let taskItem: TaskViewModel | null = null;
          if (task.action === CardTaskAction.Complete
            && (taskItem = taskItems.find(x => x instanceof TaskViewModel && x.taskModel.cardTask!.rowId === task.rowId)!)
          ) {
            this.validateFromRules(taskItem.taskModel, failRules);
          }
        }
      }
    }
  }

  private validateFromRules(cardModel: ICardModel, failRules: KrPermissionMandatoryRuleStorage[]) {
    for (let controlViewModel of cardModel.controlsBag) {
      const sourceInfo = controlViewModel.cardTypeControl.getSourceInfo();
      if (sourceInfo
        && failRules.some(x => x.sectionId === sourceInfo.sectionId
            && (x.columnIds.length === 0 || sourceInfo.columnIds.some(y => x.columnIds.some(z => z === y))))
      ) {
        controlViewModel.validationFunc = (c) => {
          if (c.hasEmptyValue) {
            return LocalizationManager.instance.format('$KrPermissions_MandatoryControlTemplate', c.caption);
          }

          return null;
        };
        controlViewModel.notifyUpdateValidation();
      }
    }
  }

}

File: Source/Extensions/Tessa.Extensions.Default.Client/Cards/KrPermissionsMandatoryStoreExtension.cs
using System.Collections;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using Tessa.Cards;
using Tessa.Cards.Extensions;
using Tessa.Extensions.Default.Shared.Workflow.KrPermissions;
using Tessa.Localization;
using Tessa.Platform.Collections;
using Tessa.UI;
using Tessa.UI.Cards;
using Tessa.UI.Cards.Tasks;

namespace Tessa.Extensions.Default.Client.Cards
{
    public sealed class KrPermissionsMandatoryStoreExtension : CardStoreExtension
    {
        #region Constructors

        public KrPermissionsMandatoryStoreExtension(ICardControlAdditionalInfoRegistry controlRegistry) =>
            this.controlRegistry = NotNullOrThrow(controlRegistry);

        #endregion

        #region Fields

        private readonly ICardControlAdditionalInfoRegistry controlRegistry;

        #endregion

        #region Base Overrides

        public override async Task AfterRequest(ICardStoreExtensionContext context)
        {
            if (!context.RequestIsSuccessful
                && UIContext.Current.CardEditor != null
                && context.Response.Info.TryGetValue(KrPermissionsHelper.FailedMandatoryRulesKey, out var rulesObj)
                && rulesObj is IList rules)
            {
                var cardModel = UIContext.Current.CardEditor.CardModel;
                var failRules = rules.Cast<Dictionary<string, object>>().Select(x => new KrPermissionMandatoryRuleStorage(x)).ToList();

                ValidateFromRules(cardModel, failRules);

                var requestTasks = context.Request.Card.Tasks;
                var taskItems = cardModel.TryGetTaskItems();
                if (requestTasks.Count > 0
                    && taskItems is { Count: > 0 })
                {
                    foreach (var task in requestTasks)
                    {
                        if (task.Action == CardTaskAction.Complete
                            && taskItems.TryFirst(x => (x as TaskViewModel)?.TaskModel.CardTask.RowID == task.RowID, out var taskItem))
                        {
                            ValidateFromRules(((TaskViewModel) taskItem).TaskModel, failRules);
                        }
                    }
                }
            }
        }

        #endregion

        #region Private Methods

        private void ValidateFromRules(ICardModel cardModel, List<KrPermissionMandatoryRuleStorage> failRules)
        {
            foreach (var controlViewModel in cardModel.ControlBag)
            {
                var sourceInfo = this.controlRegistry.GetSourceInfo(controlViewModel.CardTypeControl);
                if (sourceInfo != null
                    && failRules.Any(x =>
                        x.SectionID == sourceInfo.SectionID
                        && (x.ColumnIDs.Count == 0
                            || sourceInfo.ColumnIDs.Any(y => x.ColumnIDs.Contains(y)))))
                {
                    controlViewModel.ValidationFunc = c => c.HasEmptyValue()
                        ? LocalizationManager.Format("$KrPermissions_MandatoryControlTemplate", controlViewModel.Caption)
                        : null;
                    controlViewModel.NotifyUpdateValidation();
                }
            }
        }

        #endregion
    }
}
